### ISSUE TITLE
A null pointer dereference in jp2_decode in jp2_dec.c  

### ISSUE THREAD

**Issue by chibataiki** [jasper-software/jasper#269]
Hello,
  There is a Null Pointer Dereference  in  master version 2.026 , 5083b94 .

poc :
[jasper_poc_v2.026.zip](https://github.com/jasper-software/jasper/files/6127314/jasper_poc_v2.026.zip)

reproduct:
./jasper --output 1.jpg --input   [poc]
asan report:

```
833531==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000048 (pc 0x7f195f9ef25f bp 0x7fffe21f4730 sp 0x7fffe21f4540 T0)
==833531==The signal is caused by a WRITE memory access.
==833531==Hint: address points to the zero page.
    #0 0x7f195f9ef25e in jp2_decode /home//jasper/src/libjasper/jp2/jp2_dec.c:454
    #1 0x7f195f9d0641 in jas_image_decode /home//jasper/src/libjasper/base/jas_image.c:436
    #2 0x556af3f2cee1 in main /home//jasper/src/appl/jasper.c:235
    #3 0x7f195f7cf0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #4 0x556af3f2c6ad in _start (/home//jasper/install_dir_check/bin/jasper+0x46ad)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /home//jasper/src/libjasper/jp2/jp2_dec.c:454 in jp2_decode
==833531==ABORTING
```

gdb : 
```
─ source:/home/goushi/jo[...].c+454 ────
449	 			  dec->cdef->data.cdef.ents[i].type,
    450	 			  dec->cdef->data.cdef.ents[i].assoc));
    451	 		}
    452	 	} else {
    453	 		for (i = 0; i < dec->numchans; ++i) {
            // image=0x00007fffffffd860  →  0x0000000000000000, dec=0x00007fffffffd858  →  [...]  →  0x0000000000000000, i=0x2
 →  454	 			jas_image_setcmpttype(dec->image, dec->chantocmptlut[i],
    455	 			  jp2_getct(jas_image_clrspc(dec->image), 0, i + 1));
    456	 		}
    457	 	}
    458
    459	 	/* Delete any components that are not of interest.
── trace ────
[#0] 0x7ffff74c625f → jp2_decode(in=0x60b0000000f0, optstr=0x0)
[#1] 0x7ffff74a7642 → jas_image_decode(in=0x60b0000000f0, fmt=0x3, optstr=0x0)
[#2] 0x555555558ee2 → main(argc=0x5, argv=0x7fffffffdd18)
```

**Comment by thoger** [comment]
Note that this closely related to / the same issue as #265.  That issue was closed as resolved via the patch for #264, but it apparently only resolved the problem by making it unreachable with the specific test file.

As previously mentioned in #265, the crash was first reproducible in version 2.0.20, specifically as of commit a4dc77c.  The relationship of that commit to this crash wasn't obvious to me, so here's what I found investigating it further.

The relevant part of the a4dc77c is the `depth` check added to `jas_image_cmpt_create()` - both test images, here and in #265, fail the check because of both `depth` and `sgnd` being 1.  Due to that, `jas_image_cmpt_create()` returns NULL to its caller `jas_image_addcmpt()`, which immediately reports error with -1 return value:

https://github.com/jasper-software/jasper/blob/version-2.0.27/src/libjasper/base/jas_image.c#L858-L863

This error condition is handled in `jas_image_depalettize()` and causes error to be returned by that function as well:

https://github.com/jasper-software/jasper/blob/version-2.0.27/src/libjasper/base/jas_image.c#L1043-L1045

However, the error return from `jas_image_depalettize()` is not handled in `jp2_decode()`:

https://github.com/jasper-software/jasper/blob/version-2.0.27/src/libjasper/jp2/jp2_dec.c#L389-L393

In that code, `dec->chantocmptlut[channo]` is set to refer to component that wasn't created, resulting in an attempt to use an uninitialized memory here:

https://github.com/jasper-software/jasper/blob/version-2.0.26/src/libjasper/jp2/jp2_dec.c#L454-L455

The fix in f94e749 should guard against uninitialized memory use at that spot, but I wonder if this should be / have been fixed via proper check if `jas_image_depalettize()` failed.

I actually see that #268 /  c4144a6 had the same cause, it was only triggering the uninitialized memory use at a different place:

https://github.com/jasper-software/jasper/blob/version-2.0.25/src/libjasper/jp2/jp2_dec.c#L438-L439

**Comment by thoger** [comment]
Note that I do not consider a4dc77c commit to introduce the issue or 2.0.20 be the first affected version.  There were other reasons why `jas_image_cmpt_create()` could return error prior to that commit.  However, is seems that the reproducers triggering the `depth` check failures are easier to create via fuzzing.

**Comment by mdadams** [comment]
@thoger I hope that I understood your comments correctly.  I agree with your assessment that the return type of the depalettizing function should be checked and also there is a null pointer dereference at least in some cases.  This should be fixed now in commit 717aeda40883b91f51161ea6c5dcac5f50ef7a59.  Can you please confirm that I did understand your comments correctly and things seem okay now on your side?  Thanks.
